lib/internal/modules/cjs/loader.js

```diff
@@ -1091,6 +1091,22 @@ Module.prototype.require = function(id) {
   requireDepth++;
   try {
     return Module._load(id, this, /* isMain */ false);
+  } catch (e) {
+    if (e.code === 'MODULE_NOT_FOUND') {
+      // A module might be created synchronously in which case Node might
+      // errornously not find it - so try again without the statCache
+      const oldStatCache = statCache;
+      statCache = new Map();
+      try {
+        return Module._load(id, this, /* isMain */ false);
```

Comment:
Shouldn't `statCache` be merged with `oldStatCache` when it succeeds? Otherwise any stat cache entries resulting from `require()`s inside the module on disk will get thrown away.

Reply:
The assumption I made is that if the second attempt succeeded someone "touched the filesystem" which potentially invalidates a lot of requests. I would prefer future requests to perform `stat` calls rather than perform a retry without the cache each time (and merge).

I am not sure this is the right assumption to make - if you think the case where throwing away the statCache hurts performance is more common in the real world I can optimise for that instead.

It would be easy to make a benchmark that optimises for either assumption and when I tested it in two "real world" code bases I have access to I didn't get a meaningful difference between the two.

For reference, here is a diff if you want to test a version that merges on success:
```
diff --git a/lib/internal/modules/cjs/loader.js b/lib/internal/modules/cjs/loader.js
index 3f4c37b6c1..0ca37bfd87 100644
--- a/lib/internal/modules/cjs/loader.js
+++ b/lib/internal/modules/cjs/loader.js
@@ -1098,7 +1098,9 @@ Module.prototype.require = function(id) {
       const oldStatCache = statCache;
       statCache = new Map();
       try {
-        return Module._load(id, this, /* isMain */ false);
+        const loadResult = Module._load(id, this, /* isMain */ false);
+        statCache = new Map([...oldStatCache, ...statCache])
+        return loadResult;
       } catch (e) {
         // If it failed, nothing was wrong with the statCache and this was
         // just a conditional require so restore it
```